src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,658 @@
+ *      try {
+ *          std::unique_ptr<streambuf> my_buffer(new streambuf);
+ *      }
+ *      catch(const std::bad_alloc &) {
+ *          print_err("Cannot allocate memory for input buffer!\n");
+ *          return -1;
+ *      }
+ *      dio::istream input_b(std::move(my_buffer)); // Takes ownership of my_buffer
+ *
+ *      // Construction with a file descriptor number
+ *      dio::istream input_c(3);
+ *      if (input_c.buffer_failure()) return -1; // Failure in buffer allocation
+ *
+ *      // Construction with a combination of a file descriptor and moving a
+ *      // pre-allocated unique_ptr of the buffer
+ *      dio::istream input_d(3, std::move(buffer_2));
+ *
+ * Also, setting (or changing) file descriptor after construction is available through set_fd() function
+ * (See note in top of set_fd()). opening and closing files is possible through open() function.
+ *
+ * Copy/move constructors/assignments and destructor
+ * -------------------------------------------------
+ *
+ * istream objects cannot be moved, copied or assigned.
+ *
+ * The istream destructer will close its fd.
+ *
+ * Public member inherited from io_base: is_open()
+ * ------------------------------------------------------
+ *
+ * istream through base io_base class provides is_open() function which return true when current
+ * file path is open or false when it's not.
+ *
+ * Error handling and Stream states
+ * --------------------------------
+ *
+ * There are four possible states in an istream: EOF, buffer failure, string related operations failure and
+ * I/O failure. These are each represented by a bit in the value returned by the current_state() function.
+ *
+ * 1. eofbit is false by default and means there is nothing left in the file descriptor and the buffer
+ *    of the stream.
+ * 2. buffer_failbit is false by default and means the class constructor failed to allocate a unique_ptr
+ *    of the buffer at construction time.
+ * 3. string_failbit is false by default and means a std::string related operation was failed.
+ * 4. io_failbit is false by default and means there is something wrong from system I/O calls.
+ *
+ * Note: if the io_failbit is set in the return value of current_state(), then io_failure() will
+ * return the errno value corresponding to the failed operation. If it is clear, io_failure() will return 0.
+ *
+ * Exception handling
+ * ------------------
+ *
+ * See the description for each function to know what exceptions can be thrown by functions.
+ *
+ * All of the exception-throwing functions have exception-free variants which are marked by the "_nx"
+ * suffix. Those functions guarantee not to throw any exceptions.
+ */
+class istream : public io_base
+{
+    // Variables to capture stream current status:
+    bool eof_state = false;
+    bool string_failed = false;
+    // io_error has different functionality, it will be set to errno which returned by system
```

Comment:
different functionality ... to what?? Does this even need to be said?

Reply:
Yeah, you're right, it doesn't.